components/notebook-controller/controllers/culling_controller.go

```diff
@@ -0,0 +1,459 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/kubeflow/kubeflow/components/notebook-controller/api/v1beta1"
+	"github.com/kubeflow/kubeflow/components/notebook-controller/pkg/metrics"
+)
+
+// The constants with name 'DEFAULT_{ENV_Var}' are the default values to be
+// used, if the respective ENV vars are not present.
+// All the time numbers correspond to minutes.
+
+const DEFAULT_CULL_IDLE_TIME = "1440" // One day
+const DEFAULT_IDLENESS_CHECK_PERIOD = "1"
+const DEFAULT_ENABLE_CULLING = "false"
+const DEFAULT_CLUSTER_DOMAIN = "cluster.local"
+const DEFAULT_DEV = "false"
+
+var CULL_IDLE_TIME = 0
+var ENABLE_CULLING = false
+var IDLENESS_CHECK_PERIOD = 0
+var CLUSTER_DOMAIN = ""
+var DEV = false
+
+// When a Resource should be stopped/culled, then the controller should add this
+// annotation in the Resource's Metadata. Then, inside the reconcile loop,
+// the controller must check if this annotation is set and then apply the
+// respective culling logic for that Resource. The value of the annotation will
+// be a timestamp of when the Resource was stopped/culled.
+//
+// In case of Notebooks, the controller will reduce the replicas to 0 if
+// this annotation is set. If it's not set, then it will make the replicas 1.
+const STOP_ANNOTATION = "kubeflow-resource-stopped"
+const LAST_ACTIVITY_ANNOTATION = "notebooks.kubeflow.org/last-activity"
+const LAST_ACTIVITY_CHECK_TIMESTAMP_ANNOTATION = "notebooks.kubeflow.org/last_activity_check_timestamp"
+
+const (
+	KERNEL_EXECUTION_STATE_IDLE     = "idle"
+	KERNEL_EXECUTION_STATE_BUSY     = "busy"
+	KERNEL_EXECUTION_STATE_STARTING = "starting"
+)
+
+// Each kernel of the Notebook Server has a status.
+// KernelStatus struct:
+type KernelStatus struct {
```

Comment:
I think we would also need to add `TerminalStatus` if it does not track terminal activity already